utils/loggingcpp/poormanprofiler.inc

```diff
@@ -0,0 +1,140 @@
+#ifndef POORMANPROFILER_
+#define POORMANPROFILER
+
+#include "stopwatch.h"
+#include <unordered_map>
+using StopWatchMapType = std::unordered_map<pthread_t, logging::StopWatch*>;
+
+void pause_(int seconds )
+{
+    struct timespec req;
+    struct timespec rem;
+    req.tv_sec  = seconds;
+    req.tv_nsec = 0;
+    rem.tv_sec  = 0;
+    rem.tv_nsec = 0;
+
+    while (1)
+    {
+        if (nanosleep(&req, &rem) != 0)
+        {
+            if (rem.tv_sec > 0 || rem.tv_nsec > 0)
+            {
+                req = rem;
+                continue;
+            }
+        }
+
+        break;
+    }
+}
+
+class PoorManProfiler
+{
+  public:
+    PoorManProfiler() {};
+    StopWatchMapType stopwatchMap;
+    pthread_mutex_t stopwatchMapMutex;
+    pthread_t timerThread;
+    bool stopwatchThreadCreated = false;
+    inline logging::StopWatch* start();
+    inline logging::StopWatch* getTimer();
+    inline void spawnAutoFinishThread();
+};
+
+void* autoFinishStopwatchThread(void* arg)
+{
+    struct timeval tvCurrent;
+
+    PoorManProfiler* profiler = reinterpret_cast<PoorManProfiler*>(arg);
+    StopWatchMapType& stopwatchMap = profiler->stopwatchMap;
+    
+    for (;;)
+    {
+        // Pause n seconds.
+        pause_(10);
+        // Iterate through the stopwatch map and see how long it's been since last activity.
+        auto stopwatchMapIter = stopwatchMap.begin();
+        logging::StopWatch* stopwatch;
+        gettimeofday(&tvCurrent, 0);
+        bool procIsIdle = true;
+
+        while (stopwatchMapIter != stopwatchMap.end())
+        {
+            stopwatch = stopwatchMapIter->second;
+
+            // If any threads have been active in the last 5 seconds, kick out.
+            if (((tvCurrent.tv_sec - stopwatch->fTvLast.tv_sec) < 5) || stopwatch->isActive())
+            {
+                procIsIdle = false;
+                break;
+            }
+
+            stopwatchMapIter++;
```

Comment:
I'm afraid there may be race condition here (i.e. in the case when inserting into the unordered_map from another thread will cause rehashing)